core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -94,199 +93,186 @@
+        int sortedCount = this.sortedCount.get();
+        int entryIndex = sortedCount == 0 ? entrySet.getHeadNextIndex() : sortedCount;
+        int nextEntryIndex = entrySet.getNextEntryIndex(entryIndex);
+        while (nextEntryIndex != NONE_NEXT) {
+            entryIndex = nextEntryIndex;
+            nextEntryIndex = entrySet.getNextEntryIndex(entryIndex);
+        }
+        return entryIndex;
     }
 
-    EntrySet.LookUp buildLookUp(int ei) {
-        return entrySet.buildLookUp(ei);
-    }
 
-    ByteBuffer readKeyFromEntryIndex(int ei) {
-        return entrySet.readKey(ei);
-    }
+    /********************************************************************************************/
+    /*-----------------------  Methods for looking up item in this chunk -----------------------*/
 
-    void setRKeyBuffer(int ei, OakDetachedReadKeyBuffer keyRef) {
-        entrySet.setKeyOutputRBuff(ei, keyRef);
-    }
+    /**
+     * look up a key
+     *
+     * @param ctx the context that will follow the operation following this key lookup.
+     *            The ctx will describe the condition of the value associated with {@code key}.
+     *            If {@code (ctx.isKeyValid() && ctx.isValueValid()) == True}, the key and value was found.
+     *            If {@code ctx.isKeyValid() == False}, there is no entry with that key in this chunk.
+     *            Otherwise, if {@code ctx.isValueValid() == False}, it means that there is an entry with that key.
+     *            It can be reused in case of {@code put}, but there is no value attached to this key.
+     *            It can happen if this entry is in the midst of being inserted, or some thread removed this key and no
+     *            rebalanace occurred leaving the entry in the chunk.
+     *            It means that the value is marked off-heap as deleted, but the connection between the entry and the
+     *            value was not unlinked yet.
+     *            {@code ctx.valueState} will indicate the reason.
+     * @param key the key to look up
+     */
+    void lookUp(ThreadContext ctx, K key) {
+        // binary search sorted part of key array to quickly find node to start search at
+        // it finds previous-to-key
+        int curr = binaryFind(ctx.tempKey, key);
+        curr = (curr == NONE_NEXT) ? entrySet.getHeadNextIndex() : entrySet.getNextEntryIndex(curr);
+
+        // iterate until end of list (or key is found)
+        while (curr != NONE_NEXT) {
+            // compare current item's key to searched key
+            boolean isValid = entrySet.keyLookUp(ctx, curr);
+            assert isValid;
+            int cmp = comparator.compareKeyAndSerializedKey(key, ctx.key.getAllocByteBuffer());
+            // if item's key is larger - we've exceeded our key
+            // it's not in chunk - no need to search further
+            if (cmp < 0) {
+                // Reset lookup key state to INVALID
+                ctx.invalidate();
```

Comment:
Can the method name explain that this is only key invalidation? Although written in the comment still confusing...

Reply:
Fixed.